### PR TITLE
Default PL100 to enable encryption in a room

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes https://github.com/matrix-org/synapse/issues/7229

TODO:
  - ~~Sytests~~ default power levels are not a spec thing
  - ~~Unit tests~~ done

### CODE REVIEW

File: tests/rest/client/test_power_levels.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        self.admin_user_id = self.register_user("admin", "pass")
+        self.admin_access_token = self.login("admin", "pass")
+        self.mod_user_id = self.register_user("mod", "pass")
+        self.mod_access_token = self.login("mod", "pass")
+        self.user_user_id = self.register_user("user", "pass")
+        self.user_access_token = self.login("user", "pass")
+
+        # Create a room
+        self.room_id = self.helper.create_room_as(
+            self.admin_access_token, tok=self.admin_access_token
```
Comment (babolivier) [inline tests/rest/client/test_power_levels.py:49]:
```suggestion
            self.admin_user_id, tok=self.admin_access_token
```

Not sure it matters much here, especially seeing as the tests pass, but let's still do it right for consistency.

File: tests/rest/client/test_power_levels.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            targ=self.mod_user_id,
+        )
+        self.helper.invite(
+            room=self.room_id,
+            src=self.admin_user_id,
+            tok=self.admin_access_token,
+            targ=self.user_user_id,
+        )
+
+        # Join the other users
```
Comment (babolivier) [inline tests/rest/client/test_power_levels.py:66]:
```suggestion
        # Make the other users join the room
```

(less double meaning)

File: tests/rest/client/test_power_levels.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        room_power_levels["users"].update({self.mod_user_id: 50})
+
+        self.helper.send_state(
+            self.room_id,
+            "m.room.power_levels",
+            room_power_levels,
+            tok=self.admin_access_token,
+        )
+
+    def test_non_admins_cannot_enable_room_encryption(self):
```
Comment (babolivier) [inline tests/rest/client/test_power_levels.py:89]: Maybe add a test that admins _can_ enable room encryption (either as a separate test or as an additional step of this one), just so we're sure it works and doesn't break in the future.

File: tests/rest/client/test_power_levels.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        # have the user try to enable room encryption
+        self.helper.send_state(
+            self.room_id,
+            "m.room.encryption",
+            {"algorithm": "m.megolm.v1.aes-sha2"},
+            tok=self.user_access_token,
+            expect_code=403,  # expect failure
+        )
+
+    def test_non_admins_cannot_send_server_acl(self):
```
Comment (babolivier) [inline tests/rest/client/test_power_levels.py:108]: Same as above.

File: tests/rest/client/test_power_levels.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            {
+                "allow": ["*"],
+                "allow_ip_literals": False,
+                "deny": ["*.evil.com", "evil.com"],
+            },
+            tok=self.user_access_token,
+            expect_code=403,  # expect failure
+        )
+
+    def test_non_admins_cannot_upgrade_room(self):
```
Comment (babolivier) [inline tests/rest/client/test_power_levels.py:135]: Same as above.

Comment (babolivier) [reply tests/rest/client/test_power_levels.py:135]: Also:

```suggestion
    def test_non_admins_cannot_tombstone_room(self):
```

Seems a bit clearer to me wrt what operation we're actually testing here.

File: tests/rest/client/test_power_levels.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        # Create another room that will serve as our "upgraded room"
+        self.upgraded_room_id = self.helper.create_room_as(
+            self.admin_access_token, tok=self.admin_access_token
```
Comment (babolivier) [inline tests/rest/client/test_power_levels.py:138]:
```suggestion
            self.admin_user_id, tok=self.admin_access_token
```

File: tests/rest/client/v1/utils.py [hunk 7]
```diff
@@ -152,7 +161,7 @@ def send_state(self, room_id, event_type, body, tok, expect_code=200, state_key=
             path = path + "?access_token=%s" % tok
 
         request, channel = make_request(
-            self.hs.get_reactor(), "PUT", path, json.dumps(body).encode("utf8")
+            self.hs.get_reactor(), method, path, json.dumps(body).encode("utf8")
```
Comment (babolivier) [inline tests/rest/client/v1/utils.py:164]: Seems a bit weird to me that we're trying to send a body even when we're doing a GET request.

File: changelog.d/7230.feature [hunk 8]
```diff
@@ -0,0 +1 @@
+Change defaults to require power level 100 for enabling encryption in a room.
```
Comment (babolivier) [inline changelog.d/7230.feature:1]: More of a nitpick here, so feel free to ignore, but something along the lines of "Change the default power level required to enable encryption in a room to 100." feels slightly simpler to read imho (and directly describes the consequence to the user).

File: tests/rest/client/test_power_levels.py (hunk 3 above)
Comment (anoadragon453) [reply tests/rest/client/test_power_levels.py:89]: I didn't test the positive cases here as those are covered in sytest. Do you think they should be added to the unit tests as well?

File: changelog.d/7230.feature (hunk 8 above)
Comment (anoadragon453) [reply changelog.d/7230.feature:1]: I might even go farther and not mention power levels, and just say "Admin" which users may be more familiar with. What do you think?

Comment (babolivier) [reply changelog.d/7230.feature:1]: wfm

File: tests/rest/client/test_power_levels.py (hunk 3 above)
Comment (babolivier) [reply tests/rest/client/test_power_levels.py:89]: I tend to think that more test coverage is always a good thing, but it's true that some sytests have the side effect to ensure these events can be sent by the room creator, who should be an admin. So *shrug*.

Comment (anoadragon453) [reply tests/rest/client/test_power_levels.py:89]: I've added positive test cases, it was simple :)

**Comment by anoadragon453** [comment]
One last review-request for the positive test cases :)

File: tests/rest/client/v1/utils.py [hunk 9]
```diff
@@ -152,8 +161,12 @@ def send_state(self, room_id, event_type, body, tok, expect_code=200, state_key=
             path = path + "?access_token=%s" % tok
 
         request, channel = make_request(
-            self.hs.get_reactor(), "PUT", path, json.dumps(body).encode("utf8")
+            self.hs.get_reactor(),
+            method,
+            path,
+            content=json.dumps(body).encode("utf8") if body else b"",
```
Comment (babolivier) [inline tests/rest/client/v1/utils.py:167]: This fails if `body` is `None`. I suggest:

```python
content = b""
if body is not None:
    content = json.dumps(body).encode("utf8")

request, channel = make_request(
    self.hs.get_reactor(), method, path, content
)
```

Comment (anoadragon453) [reply tests/rest/client/v1/utils.py:167]: I think it's failing because `body` being `{}` is falsey?

Comment (anoadragon453) [reply tests/rest/client/v1/utils.py]: Your code works though :)

**Pull request merged**